core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlAstBuilder.scala

```diff
@@ -12,8 +11,10 @@ import scala.collection.JavaConverters.asScalaBufferConverter
  */
 class TSqlAstBuilder extends TSqlParserBaseVisitor[ir.LogicalPlan] {
 
-  private val relationBuilder = new TSqlRelationBuilder
-  private val expressionBuilder = new TSqlExpressionBuilder
+  private val expressionBuilder = new TSqlExpressionBuilder(null)
+  private val relationBuilder = new TSqlRelationBuilder(expressionBuilder)
+  expressionBuilder.relationBuilder = relationBuilder
```

Comment:
if expression builder and relation builder need each other - they have to be in the same class. I guess `TSqlExpressionBuilder` :)

Reply:
Might be that the visitors that cause this interplay need moving to be builders. Moving to injection here only points out what the relationship already was of course.

Reply:
I implemented a visitor coordinator such that each visitor has access to the other visitors if it needs them, without being dependent on any particular implementation of the other visitors. The VisitorCoordinator is then injected into each visitor. The coordinator can also supply any other shared resources using the same technique.